scripts/utils/consensus-celltype-plotting-functions.R

```diff
@@ -0,0 +1,185 @@
+    # groups similar cell types together 
+    dplyr::left_join(validation_groups_dt, by = "consensus_annotation") |> 
+    dplyr::left_join(gene_exp_dt, by = c("barcodes", "library_id")) |> 
+    # add marker gene information (associated validation group annotation, gene observed count, percent tissues)
+    # account for the same gene being present in multiple cell types 
+    dplyr::left_join(markers_dt, by = "ensembl_gene_id", relationship = "many-to-many") |> 
+    dplyr::mutate(detected = logcounts > 0)
+  
+  # remove extra data frame
+  rm(gene_exp_dt)
+  gc()
+  
+  # prep for plots 
+  # get total number of cells per final annotation group 
+  total_cells_df <- consensus_dt |> 
+    dplyr::select(library_id, barcodes, broad_celltype_group) |> 
+    dplyr::distinct() |> 
+    dplyr::count(broad_celltype_group, name = "total_cells") 
+  
+  # table with one row per unique broad cell type/ marker gene combination 
+  # first all cells in with the same broad_celltype_group (determined based on consensus_annotation) are grouped together
+  # then get the mean gene expression and total percentage of cells that express each marker gene across all cells in that group
+  # do this for every possible marker gene/ validation group assignment 
+  # second we calculate the mean expression and mean percentage of all marker genes in a given validation group (this value is used only in the second section of the report)
+  group_stats_df <- consensus_dt |> 
+    # for each assigned cell type/marker gene combo get total detected and mean expression
+    # group by both broad group and validation group to account for genes that are expressed in more than one cell type
+    dplyr::group_by(broad_celltype_group, ensembl_gene_id, validation_group_annotation) |>
+    dplyr::summarize(
+      detected_count = sum(detected),
+      mean_exp = mean(logcounts)
+    ) |> 
+    # add in validation group for marker genes
+    # this includes all possible marker genes and all possible validation group assignments 
+    dplyr::left_join(markers_dt, by = c("ensembl_gene_id", "validation_group_annotation"), relationship = "many-to-many") |>
+    # now get the mean expression/ mean percentage across all marker genes for a given validation group
+    # here the broad_celltype_group is the final assigned annotation for that group of cells 
+    # the validation_group_annotation refers to the cell type that marker gene is associated with 
+    dplyr::group_by(broad_celltype_group, validation_group_annotation) |> 
+    dplyr::mutate(
+      # calculate mean expression/detected across all markers for a specific group 
+      all_markers_mean_exp = mean(mean_exp),
+      all_markers_detected_count = mean(detected_count)
+    ) |>  # add total cells
+    dplyr::left_join(total_cells_df, by = c("broad_celltype_group")) |> 
+    # for plotting we're only going to look at any cell types with > 50 cells otherwise these plots can get wild 
+    dplyr::filter(total_cells > 50) |> 
```

Comment:
I think make this a parameter, in case you want to vary it? In your comments you suggest that 200 was another version you played with; can we see both to compare?